### DISCUSSION TITLE
Is it possible to profile mmap memory leak?

### DISCUSSION THREAD

**Discussion by roy-20230331** [async-profiler/async-profiler#980] [Q&A] [answered]
Dear @apangin , I found we have experimental feature to find malloc memory leaks(https://github.com/async-profiler/async-profiler/tree/malloc#finding-native-memory-leaks), is it possible to merge into production?

BTW, is it possible to provide mmap memory leak profiling feature as well?

If we don't have this feature, do you have any recommendation that I can troubleshooting mmap memory leaks? Recently we use `mmap` to load MaxMind GeoIP data into off-heap memory, but have off-heap memory leak issue...

Really appreciate ur great help! 

Thanks,
Roy

**Comment by apangin** [comment, marked as answer]
Why do you think it's a mmap leak? (or what do you call "mmap leak", to begin with?)

I assume you've seen [this presentation](https://www.youtube.com/watch?v=c755fFv1Rnk)?
You may start analyzing with `pmap -x PID`. This gives you a memory map of the process. Take multiple snapshots over time and find difference: which regions are growing.

Regarding timeline of the native memory profiling feature, I answered [here](https://github.com/async-profiler/async-profiler/discussions/836#discussioncomment-10055271).

**Comment by roy-20230331** [comment]
>> Why do you think it's a mmap leak? 

After run for several hours, linux killed our application. We find "kernel: out_of_memory" in dmesg.

>> I assume you've seen [this presentation](https://www.youtube.com/watch?v=c755fFv1Rnk)?

Absolutely! I learnt a lot from this presentation, inspired me a lot. Thank you @apangin!

>> You may start analyzing with pmap -x PID. This gives you a memory map of the process. Take multiple snapshots over time and find difference: which regions are growing.

Sure, let me begin with pmap, thanks again @apangin !
